OSDOCS-7640: Doc how service accounts assume AWS IAM roles (#64331)
openshift/openshift-docs: modules/how-service-accounts-assume-aws-iam-roles-in-default-projects.adoc

Cluster Operators use service accounts to assume IAM roles. When a service account assumes an IAM role, temporary STS credentials are provided for the service account to use in the cluster Operator's pod. If the assumed role has the necessary AWS privileges, the service account can run AWS SDK operations in the pod.

[discrete]
[id="workflow-for-assuming-aws-iam-roles-in-default-projects_{context}"]
== Workflow for assuming AWS IAM roles in default projects

// The following diagram illustrates the workflow by which the service accounts for cluster Operators assume IAM roles:
//
// .Workflow for assuming AWS IAM roles in default projects 
// TODO: Add diagram.

The workflow has the following stages:

. Within each project that a cluster Operator runs, a deployment spec defines a service account token configuration. These tokens are audience- and time-bound. Every hour, ROSA generates a new token, which generates the following temporary credentials mounted to the cluster Operator's pod:

> ** An `$AWS_ARN_ROLE` environment variable that contains the ARN for the IAM role that has the permissions required to run AWS SDK operations.

Reviewer: ```suggestion
** An `$AWS_ARN_ROLE` variable that contains the ARN for the IAM role that has the permissions required to run AWS SDK operations.
```

> ** An `$AWS_WEB_IDENTITY_TOKEN_FILE` environment variable that contains the full path in the pod to the OpenID Connect (OIDC) token for the service account. The full path is `/var/run/secrets/eks.amazonaws.com/serviceaccount/token`.

Reviewer: ```suggestion
** An `$AWS_WEB_IDENTITY_TOKEN_FILE` variable that contains the full path in the pod to the OpenID Connect (OIDC) token for the service account. The full path is `/var/run/secrets/openshift/serviceaccount/token`.
```

Reviewer: Is this for OpenShift cluster operators or the pod identity webhook? If its for the pod identity webhook then ignore this suggestion as the eks path is correct.

Author: This is for the cluster Operators, so I corrected the path as suggested.

> . When a cluster Operator needs to assume an AWS IAM role to access an AWS service (such as EC2), the AWS SDK client code running on the Operator invokes the `AssumeRoleWithWebIdentity()` API call.

Reviewer: ```suggestion
. When a cluster Operator needs to assume an AWS IAM role to access an AWS service (such as EC2), the AWS SDK client code running on the Operator invokes the `AssumeRoleWithWebIdentity` API call.
```

. The OIDC token is passed from the pod to the OIDC provider. The provider authenticates the service account identity if the following requirements are met:

** The identity signature is valid and signed by the private key.

** The `sts.amazonaws.com` audience is listed in the OIDC token and matches the audience configured in the OIDC provider.
+
[NOTE]
====
In {product-title} with STS clusters, the OIDC provider is created during install and set as the service account issuer by default. The `sts.amazonaws.com` audience is set by default in the OIDC provider.
====

** The OIDC token has not expired.

** The issuer value in the token contains the URL for the OIDC provider.

. If the project and service account are in the scope of the trust policy for the IAM role that is being assumed, then authorization succeeds.

> . After successful authentication and authorization, temporary AWS STS credentials in the form of a session token are passed to the pod for use by the service account. By using the credentials, the service account is temporarily granted the AWS permissions enabled in the IAM role.

Reviewer: ```suggestion
. After successful authentication and authorization, temporary AWS STS credentials in the form of a AWS access key, secret key and session token are passed to the pod for use by the service account. By using the credentials, the service account is temporarily granted the AWS permissions enabled in the IAM role.
```

> . When the cluster Operator runs AWS SDK operations in the pod, the service account provides the temporary STS credentials to the AWS API to verify its identity.

Reviewer: ```suggestion
. When the cluster Operator runs the operator using the AWS SDK in the pod consumes the secret containing the path to the projected service account and AWS IAM Role ARN to authenticate against the OIDC provider which returns temporary STS credentials for authentication against the AWS API.
```
